src/test/ui/generator/auto-trait-regions.rs

```diff
@@ -18,36 +24,42 @@ impl Foo for &'static OnlyFooIfStaticRef {}
 struct OnlyFooIfRef(No);
 impl<'a> Foo for &'a OnlyFooIfRef {}
 
-fn assert_foo<T: Foo>(f: T) {}
+fn assert_foo<T: Foo>(_f: T) {}
 
 fn main() {
-    // Make sure 'static is erased for generator interiors so we can't match it in trait selection
-    let x: &'static _ = &OnlyFooIfStaticRef(No);
+    let z = OnlyFooIfStaticRef(No);
+    let x = &z;
+    //~^ ERROR `z` does not live long enough
     let gen = || {
         let x = x;
         yield;
         assert_foo(x);
     };
-    assert_foo(gen);
-    //~^ ERROR implementation of `Foo` is not general enough
-    //~| ERROR implementation of `Foo` is not general enough
+    assert_foo(gen); // bad
+    drop(z);
+    //~^ ERROR cannot move out of `z` because it is borrowed
 
     // Allow impls which matches any lifetime
-    let x = &OnlyFooIfRef(No);
+    let z = OnlyFooIfRef(No);
+    let x = &z;
     let gen = || {
         let x = x;
         yield;
         assert_foo(x);
     };
     assert_foo(gen); // ok
+    drop(z);
 
-    // Disallow impls which relates lifetimes in the generator interior
-    let gen = || {
-        let a = A(&mut true, &mut true, No);
+    let gen = static || {
+        let mut y = true;
+        let a = A::<'static, '_>(Cell::new(&true), Cell::new(&mut y), No);
         yield;
-        assert_foo(a);
+        drop(a);
     };
-    assert_foo(gen);
-    //~^ ERROR not general enough
-    //~| ERROR not general enough
+    assert_foo(gen); // bad -- why does this pass?
```

Comment:
So I was thinking about this test case recently. 

I think why this doesn't fail is because we don't actually enforce any dropck... So there's nothing enforcing the lifetime needs to be shorter than `'static`.

Wonder if there's any way of hooking that information in during auto trait confirmation.

Reply:
If I'm understanding this right, this seems concerning. So the issue is that `a` is `!Foo` because of the `No` field included in the struct. Then `a` obviously lives across the `yield` point since we explicitly drop it afterwards. This means `a` needs to be saved in the generator, so the resulting generator should be `!Foo` as well, but yet `assert_foo(gen)` passes...

Oh, but then we explicitly impl `Foo` for `A`, which overrides the auto rules, right? However, the impl shouldn't apply because the `Cell::new(&mut y)` field is not bound by `'static` since `y` ends at the end of the generator.

I agree with you, it seems like this shouldn't pass, and it seems likely to permit undefined behavior if it does...

Maybe the compiler is doing something smart elsewhere by recognizing that `a` is destructed at the end of the generator as well so it can automatically shorten the lifetimes? That seems like that'd be extremely difficult to do safely though, so that seems unlikely to me.

Reply:
Yeah, this is because of the fact we refresh the late bound regions with totally new inference variables. While stashing away all of the region constraints we collect in generator_interior is okay to enforce _most_ borrowchecky things, we still can't enforce anything like requiring one local variable living for longer than another (or static) in the generator interior with my implementation as-is. 

But yeah, this one's the case that I need someone who knows borrowck/dropck better to help with.